Modules/_lsprof.c

```diff
@@ -650,6 +594,99 @@ setBuiltins(ProfilerObject *pObj, int nvalue)
     return 0;
 }
 
+PyObject* pystart_callback(ProfilerObject* self, PyObject *const *args, Py_ssize_t size)
+{
+    PyObject* code = args[0];
+    ptrace_enter_call((PyObject*)self, (void *)code, (PyObject *)code);
+
+    Py_RETURN_NONE;
+}
+
+PyObject* pyreturn_callback(ProfilerObject* self, PyObject *const *args, Py_ssize_t size)
+{
+    PyObject* code = args[0];
+    ptrace_leave_call((PyObject*)self, (void *)code);
+
+    Py_RETURN_NONE;
+}
+
+PyObject* get_cfunc_from_callable(PyObject* callable, PyObject* self_arg, PyObject* missing)
+{
+    // return a new reference
+    if (PyCFunction_Check(callable)) {
+        Py_INCREF(callable);
+        return (PyObject*)((PyCFunctionObject *)callable);
+    }
+    if (Py_TYPE(callable) == &PyMethodDescr_Type) {
```

Comment:
Is this necessary?
Doesn't he profiler extracts the same data from the builtin function that it could from the method descriptor?

Reply:
This piece is copied from the new `setprofile` I believe. The idea behind it is to make sure `get_cfunc_from_callable` only returns a `PyCFunctionObject`. If it's not, then `((PyCFunctionObject *)cfunc)->m_ml` won't work.

What do you propose here? Simply return `callable`? We need to check that anyway because `CALL` event can be triggered before calling a Python function and we don't want to add profiler entry on that(it should be dealt with later). I did realize that `Py_RETURN_NONE` was incorrect - `NULL` should be returned.

Reply:
All that happens to the builtin function objects is it gets passed down to `normalizeUserObj()` which then does some elaborate lookup to get the method descriptor back again. Both the method descriptor and builtin function contain a pointer to the same `PyMethodDef` struct.

So, leave the method descriptor alone here, and in `normalizeUserObj()` create the same string that as would be created for the builtin function.

Reply:
The `mo` object [on line 175](https://github.com/python/cpython/pull/103534/files#diff-fe854c1e90974a430c6672f28fcf6e470b6be183d515b851367581fed030b6c3R175) is the method descriptor.

Reply:
There are three purposes `get_cfunc_from_callable` needs the serve:

1. To get the object(callable) to print later, which could be resolved in `normalizeUserObj()` like you said.
2. To get a unique key for the hash table(actually a binary tree). In this case, `((PyCFunctionObject *)cfunc)->m_ml` was used, maybe the callable itself could be used as well? Not sure why `m_ml` was chosen at the first place, the builtin functions probably have dinstinct addresses anyway? Even for the builtin methods(like `list.append`), using `(void*)callable` directly might work?
3. To filter out unwanted data. This is why this part is necessary. Without trying to get the C function from the method, how could we know if this is a call to a builtin function that we would like to log? It could be a simple Python call(`f()`), which would trigger `CALL` event. It could be a method call, but not to a C function(`self.some_method()`). We need to filter these entry out before we even reach `normalizeUserObj()`(that's where we are about to log the function call).

The current implementation is trying to swap out the `setprofile` layer without touching the internal profiling system. It's true that the profiling system could be optimized, but it's also risky and probably need extra care. If we want to land this in 3.12, maybe we should avoid changing the profiling logic for now.

Reply:
I agree with your point about risk.

I think it is worth cleaning up the internals. We couldn't do it before, as the conversion from method descriptor to builtin function occurred before cprofile got to see it. Maybe we can get it done for 3.12, maybe not.

So let's get this change in for 3.12, and we can streamline things later.

Reply:
We can clean up the internals, but we still need to address point 3. We need to filter out entries that we don't want, and that probably requires resolving the descriptors.

Reply:
All the information in the fake builtin function is also in the original method descriptor, so whatever the filter was doing should still work.

Reply:
Do you suggest that we can get rid of `Py_TYPE(callable)->tp_descr_get(callable, self_arg, (PyObject*)Py_TYPE(self_arg))` because we don't need the actual `PyCFunctionObject` in it? We still need to check against `Py_TYPE(callable) == &PyMethodDescr_Type` for the actual "builtin methods" right? To filter out Python defined methods?